### DISCUSSION TITLE
Standalone Language Server

### DISCUSSION THREAD

**Discussion by simeonoff** [wkillerud/some-sass#111] [Show and tell]
First and foremost, I can't express how grateful I am for all your hard work on this project. This makes Sass feel closer to a grown-boy language with all niceties one may expect when working with it in a text editor.

Now with that out of the way, I use Neovim as my editor of choice and wanted to get the language server capabilities to work there too. For that reason I created [Some Sass Language Server Extracted](https://github.com/simeonoff/some-sass-langserver-extracted) that aims to extract and build the language server from this repo and then publish it on npm as a standalone, installable binary that can easily be integrated in any LSP-capable editor.

I hope you don't mind me doing so!

**Comment by wkillerud** [comment]
That's awesome @simeonoff, I love it! 

I've had something like this on my "Maybe some day" list for a while (inspired by [vscode-languageservers-extracted](https://github.com/hrsh7th/vscode-langservers-extracted)), looks like you beat me to it 😄 


I use [semantic release](https://github.com/wkillerud/vscode-scss/blob/main/release.config.cjs) to publish the extension itself. If you want to, I'd be more than happy to collaborate on setting something up here in this repo that autopublishes updates to the server (with your additions) to npm.

**Reply by simeonoff** [reply]
That would be awesome. I was also thinking of having the server published on the [Mason Registry](https://mason-registry.dev/) for easier installation via Mason in Neovim. Still gotta read the contributing guidelines, but I believe I'm halfway there. Additionally, the base server setup should be added to the [mason-lspconfig](https://github.com/williamboman/mason-lspconfig.nvim) repo as well.

Alright, lets connect so we can automate the release of the server. I will drop you an email and we can take it from there.

**Reply by wkillerud** [reply]
I opened a draft PR https://github.com/wkillerud/vscode-scss/pull/112 which we can use as a starting point, but feel free to shoot me an email 😄 

I suggest we don't package and publish the language server for other editors here in this repo. Instead I imagine you'd set up a new repo for the Mason Registry stuff and something like [renovate](https://github.com/renovatebot/renovate#readme) to get PRs whenever there's a new version on npm. With [automerge](https://docs.renovatebot.com/presets-default/#automergeminor) and GitHub Actions the whole process could be automated for non-breaking changes.

**Reply by simeonoff** [reply]
That was quick! I don't believe I need a separate repo for mason as it will pull the binary straight from the published package on npm. I already created a [PR](https://github.com/mason-org/mason-registry/pull/4746) that adds the package.

I will look into updating my repo with what you've set up here as it eliminates the need to do half the things I've done. I guess my repo will just be a proxy for this one that simply plucks out the server portion and pushes to npm on new releases of the extension. I will look into renovate and automerge to set up the flow tomorrow or over the weekend. I also need to write an lspconfig for mason so that the server is auto-configured upon installation.

**Reply by wkillerud** [reply]
> I don't believe I need a separate repo for mason as it will pull the binary straight from the published package on npm. I already created a [PR](https://github.com/mason-org/mason-registry/pull/4746) that adds the package.

I see. So that entry would need to be updated manually when there are updates to the language server package on npm? Does the registry support a semantic version range (`pkg:npm/some-sass-language-server@^1.0.1`), or does it have to be pinned? 

> I will look into updating my repo with what you've set up here as it eliminates the need to do half the things I've done. I guess my repo will just be a proxy for this one that simply plucks out the server portion and pushes to npm on new releases of the extension.

I think [some-sass-langserver-extracted](https://github.com/simeonoff/some-sass-langserver-extracted) can be archived once this is set up, if that's OK with you. My idea is to publish the npm package with the `bin` and everything directly so it stays up to date with the VS Code extension.

Today's situation illustrated:
![](https://github.com/wkillerud/vscode-scss/assets/1223410/2ea7e645-9bf7-4b13-b9c5-ee50e3dae5b3)

- `some-sass-langserver-extracted` pulls the source, repackages, and publishes
- Manual process or automated on a schedule
- Can get out of sync

Target:
![](https://github.com/wkillerud/vscode-scss/assets/1223410/1ab31b18-f459-4a56-ba63-79a47b5147f0)

- Binary published automatically when there are changes
- Stays in sync


What I'll need from you to make it happen is either:

- You send me a token privately so I can publish to [some-sass-language-server](https://www.npmjs.com/package/some-sass-language-server) on npm
- You invite [me](https://www.npmjs.com/~wllm) as a maintainer of the package on npm so I can generate a token myself (via the [Settings tab for the package on npm](https://www.npmjs.com/package/some-sass-language-server/access))
- You [transfer the package](https://docs.npmjs.com/transferring-a-package-from-a-user-account-to-another-user-account) to [me on npm](https://www.npmjs.com/~wllm)

Either option works for me :)

**Comment by simeonoff** [comment]
> I see. So that entry would need to be updated manually when there are updates to the language server package on npm? Does the registry support a semantic version range (`pkg:npm/some-sass-language-server@^1.0.1`), or does it have to be pinned?

The mason registry uses `renovate` to monitor npm and bump the version number upon a new release. The only requirement when creating a PR there is that a version must be specified initially, just like I did. Anytime there's any update to any of the packages in the registry, they release a new version of the mason registry with the packages bumped (scheduled). So no, no need to update it manually. As for the semantic versioning, the users of the mason registry have to consciously install any update to any of the packages listed there. I believe the process between updates to the packages and updates to the lsp configs is a bit broken, alas this is how it works there. Anytime you introduce a breaking change to the configuration of the server, the lsp config will also need to be updated. But I am more than happy to maintain that part.

> I think [some-sass-langserver-extracted](https://github.com/simeonoff/some-sass-langserver-extracted) can be archived once this is set up, if that's OK with you. My idea is to publish the npm package with the `bin` and everything directly so it stays up to date with the VS Code extension.

Got it, even better!

> What I'll need from you to make it happen is either:
> 
> * You send me a token privately so I can publish to [some-sass-language-server](https://www.npmjs.com/package/some-sass-language-server) on npm
> * You invite [me](https://www.npmjs.com/~wllm) as a maintainer of the package on npm so I can generate a token myself (via the [Settings tab for the package on npm](https://www.npmjs.com/package/some-sass-language-server/access))
> * You [transfer the package](https://docs.npmjs.com/transferring-a-package-from-a-user-account-to-another-user-account) to [me on npm](https://www.npmjs.com/~wllm)
> 
> Either option works for me :)

Let's just go with the last option as you will have full control over the package that way.

**Reply by wkillerud** [reply]
> The mason registry uses renovate to monitor npm and bump the version number upon a new release. [...] So no, no need to update it manually.

Nice, good stuff.

>  As for the semantic versioning, the users of the mason registry have to consciously install any update to any of the packages listed there.

Makes sense.

> Anytime you introduce a breaking change to the configuration of the server, the lsp config will also need to be updated. But I am more than happy to maintain that part.

That would be great 👍

> Let's just go with the last option as you will have full control over the package that way.

Thank you, I got the invite from `npm` 🙏

**Reply by wkillerud** [reply]
It wasn't as easy as I'd hoped, but it works! [v1.0.2 published automatically](https://www.npmjs.com/package/some-sass-language-server/v/1.0.2).

Thank you for setting this in motion @simeonoff 🥳 Let me know how it goes, I'd love to [list Neovim here](https://github.com/wkillerud/some-sass?tab=readme-ov-file#editors-with-clients) once the Mason Registry stuff lands 😄

**Reply by simeonoff** [reply]
Great, happy it worked! So glad to have better Sass language support in Neovim.

From what I gather, it takes some time before they merge a package on the Mason registry, but manually installing and having a manual lsp config is not too bad either for now.
